### PR TITLE
493 base layer select scale

### PR DESCRIPTION
**Pull request by badnames**
## Basics

- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildserver is happy.

## Checklist

- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [x] I fixed all affected decisions
- [x] I added unit tests for my code
- [x] I added code comments, logging, and assertions as appropriate
- [x] I translated all strings visible to the user
- [x] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [x] I created left-over issues for things that are still to be done
- [x] Code is conforming to [our Architecture](/doc/architecture)
- [x] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [x] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Review by absurd-turtle — CHANGES_REQUESTED** [review]
I think there is still a flaw in how the Auto-Scale feature is supposed to work.
When I measure half of the distance on the base layer image and click apply.
And then do the same again I get a higher scaling value.
I think the distance that is drawn should always be in relation to the image instead of to the base stage.

**Comment by absurd-turtle** [comment]
Consider getting rid of the apply button. Instead change the image size whenever the scaling value changes.

**Comment by badnames** [comment]
> I think there is still a flaw in how the Auto-Scale feature is supposed to work. When I measure half of the distance on the base layer image and click apply. And then do the same again I get a higher scaling value. I think the distance that is drawn should always be in relation to the image instead of to the base stage.

Yes I forgot too include the current scale in the calculation, but it should be fixed now.
I created this image to verify that distances are calculated correctly in the editor:
distancetest

**Comment by badnames** [comment]
> Consider getting rid of the apply button. Instead change the image size whenever the scaling value changes.
The state is now updated every time the onBlur event occurs or after scale has been calculated.

**Review by markus2330 — CHANGES_REQUESTED** [review]
I now saw the frog! There is a quite long lag between changing values and actual changes in the picture. Sometimes the auto-scaling did not work and the red points weren't where I clicked. Didn't investigate further, though.

File: frontend/package.json [hunk 1]
```diff
@@ -19,6 +19,7 @@
   "dependencies": {
     "@tanstack/react-query": "^4.29.1",
     "axios": "^1.2.3",
+    "canvas": "^2.11.2",
```
[the remark this answers is not in the document]
Comment (badnames) [reply frontend/package.json:22]: Our test environment does not run in a Browser, but in a Node.js environment that does provide the Canvas API.
This package contains the necessary definitions that make it possible to test components that rely on Konva.

File: frontend/src/features/map_planning/layers/base/BaseMeasurementLayer.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    setMeasurePoint(position);
+  };
+
+  return (
+    <Layer
+      listening={
+        untrackedState.measureStep !== 'inactive' && untrackedState.measureStep !== 'both selected'
+      }
+      width={config.width}
+      height={config.height}
```
Comment (Bushuo) [inline frontend/src/features/map_planning/layers/base/BaseMeasurementLayer.tsx:41]: This seems to be undefined.
The config is not passed as a prop from the Map.

File: frontend/src/features/map_planning/layers/base/BaseMeasurementLayer.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      onClick={measurementOnClick}
+    >
+      <Rect width={9999999} height={9999999} x={0} y={0} color="red" />
```
Comment (Bushuo) [inline frontend/src/features/map_planning/layers/base/BaseMeasurementLayer.tsx:44]: Better extract a constant here.

Is this Rect intended to catch the click events for measuring?
If yes, I think x and y need to be -LAYER_SIZE/2.
Such that the middle of the Rect is on the origin.

File: frontend/src/features/map_planning/components/Map.tsx [hunk 4]
```diff
@@ -146,15 +143,14 @@ export const Map = ({ layers }: MapProps) => {
...
-          pixelsPerMeter={trackedState.layers.base.scale}
-          rotation={trackedState.layers.base.rotation}
+          listening={selectedLayer.type_ === LayerType.Base}
         />
         <PlantsLayer
           visible={untrackedState.layers.plants.visible}
           opacity={untrackedState.layers.plants.opacity}
           listening={selectedLayer.type_ === LayerType.Plants}
         ></PlantsLayer>
+        <BaseMeasurementLayer />
```
Comment (Bushuo) [inline frontend/src/features/map_planning/components/Map.tsx:153]: I think this needs to go to the top of the list to catch the click events.

**Review by Bushuo — CHANGES_REQUESTED** [review]
Overall, nice work :)

I think the flickering has to do with how often the KonvaImage is rerendered.
This needs further investigation, but @absurd-turtle and me where refactoring it in another issue.

File: frontend/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (badnames) [reply frontend/package.json:22]: Yes! Thats a valid criticism.

File: frontend/src/features/map_planning/components/Map.tsx (hunk 4 above)
Comment (badnames) [reply frontend/src/features/map_planning/components/Map.tsx:153]: Layers on the bottom will be shown drawn on top.
Besides, I did not notice any issues regarding click events.
Did I miss anything?

**Comment by markus2330** [comment]
> Absolutely no idea. Requests which update or add files don't even work right now.

I think I found the "problem": I copied the file and it seems like that a copy gets listed in recently edited.

File: frontend/src/features/map_planning/utils/Constants.ts [hunk 5]
```diff
@@ -1 +1,2 @@
 export const MAP_PIXELS_PER_METER = 100;
+export const BASE_LAYER_MEASUREMENT_RECT_WIDTH = 999999;
```
Comment (markus2330) [inline frontend/src/features/map_planning/utils/Constants.ts:2]: Please add some documentation what this is.

File: frontend/src/features/map_planning/components/Map.tsx (hunk 4 above)
Comment (markus2330) [reply frontend/src/features/map_planning/components/Map.tsx:153]: @badnames did you try to change the order?

File: frontend/src/features/map_planning/layers/base/BaseMeasurementLayer.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import useMapStore from '@/features/map_planning/store/MapStore';
+import { BASE_LAYER_MEASUREMENT_RECT_WIDTH } from '@/features/map_planning/utils/Constants';
+import Konva from 'konva';
+import { Circle, Layer, Rect, Line } from 'react-konva';
+
+import KonvaEventObject = Konva.KonvaEventObject;
+
+/**
+ * A virtual ruler that is used to measure distances on the base layer.
```
Comment (markus2330) [inline frontend/src/features/map_planning/layers/base/BaseMeasurementLayer.tsx:9]: Is this for auto-scaling? Please comment how auto-scaling is implemented.

**Review by markus2330 — APPROVED** [review]
Much better, only the flickering is very annoying and code comments should be improved. Rebase is also needed.

**Comment by absurd-turtle** [comment]
> Much better, only the flickering is very annoying and code comments should be improved. Rebase is also needed.

The flickering could be resolved after a rebase.
We worked on that in the UI/UX improvements which got merged recently.

**Comment by absurd-turtle** [comment]
Consider adding the same element as in Plant layer which indicates that an action is currently active.
Would be useful to see during the auto scaling as well :)

**Review by absurd-turtle — DISMISSED** [review]
The previously mentioned issue with Auto-Scaling is not fixed yet.

**Comment by badnames** [comment]
> @badnames did you try to change the order?

Yes, but the only thing that changed is that the line visualizing the measured distance is behind the base layer.

**Review by Bushuo — APPROVED** [review]
Works as expected :) Well done.

**Comment by markus2330** [comment]
Works fantastic, good job! :heart:

Now we need a connection to the backend, pls check the API #584 :rocket:

**Pull request merged**